clippy_utils/src/consts.rs

```diff
@@ -606,20 +602,18 @@ impl<'a, 'tcx> ConstEvalLateContext<'a, 'tcx> {
 pub fn miri_to_const<'tcx>(tcx: TyCtxt<'tcx>, result: mir::ConstantKind<'tcx>) -> Option<Constant> {
     use rustc_middle::mir::interpret::ConstValue;
     match result {
-        mir::ConstantKind::Val(ConstValue::Scalar(Scalar::Int(int)), _) => {
-            match result.ty().kind() {
-                ty::Bool => Some(Constant::Bool(int == ScalarInt::TRUE)),
-                ty::Uint(_) | ty::Int(_) => Some(Constant::Int(int.assert_bits(int.size()))),
-                ty::Float(FloatTy::F32) => Some(Constant::F32(f32::from_bits(
-                    int.try_into().expect("invalid f32 bit representation"),
-                ))),
-                ty::Float(FloatTy::F64) => Some(Constant::F64(f64::from_bits(
-                    int.try_into().expect("invalid f64 bit representation"),
-                ))),
-                ty::RawPtr(_) => Some(Constant::RawPtr(int.assert_bits(int.size()))),
-                // FIXME: implement other conversions.
-                _ => None,
-            }
+        mir::ConstantKind::Val(ConstValue::Scalar(Scalar::Int(int)), _) => match result.ty().kind() {
+            ty::Adt(_, _) => Some(Constant::Int(int.assert_bits(int.size()))),
```

Comment:
Initially I thought of bringing the field's ident from `ExprKind::Field` here in order to search the actual type with something like `def.all_fields().any(|field| ... match ident logic ...)` but then I realized that someone probably already did this in a clever way.

The problem is, I don't know any other way. Do you have any thoughts about this matter?

Reply:
Hey @rust-lang/clippy, has someone maybe worked mir and has an idea how to solve this? :)

Reply:
My search in the deepest and darkest parts of Rustc wasn't a success and since no one else replied, I moved forward with the initial idea. Hey, at least it is working 🙄

Reply:
Would be better to add a variant to `Constant` to hold struct types and then resolve fields as they come up. The variant can just be a reference to an allocation.

Reply:
I am sorry @Jarcho, can you elaborate?

Reply:
I'm guessing he recommends adding a new variant like `Constant::Field(bits-as-int)` and let the lint handle the type of the bits inside. Does that make sense? And @Jarcho was this what you meant?

Reply:
Add something like `Constant::Adt(ConstantKind<'tcx>)`, have `Constant::fetch_path` return that when needed and then fields as needed in `Constant::expr`.

You need to be handle any nested fields (e.g. `a.b.c.d`) as well arrays and slices (can be done in a different pr).

Reply:
> Add something like `Constant::Adt(ConstantKind<'tcx>)`, have `Constant::fetch_path` return that when needed and then fields as needed in `Constant::expr`.

Done as requested